### PR TITLE
Vite migration

### PR DESCRIPTION
**Pull request by femans**
#### This PR migrates from webpack/vue-cli to vite.
It also replaces the testing package jest with vitest. (Jest does not play nice with vite, they are of different schools of js modules)
It does not (yet) update vue, vuex or vuetify. 

**N.b.** that the sass errors are still there when running `npm run serve` because vuetify has not been updated yet. This causes the build to be a bit slow. Still you should observe a pleasant speed increase of the build.

#### Migration log:
https://www.notion.so/ipfs-search/Upgrading-to-Vue3-part-1-migrating-to-Vite-cdc9fe3e553442199d53a4767cfcb4a9

#### To test this:
after `npm i`,
1. try running the code. See if all the functions of the frontend work properly. See if there are no bugs or console errors. (that were not there on the master branch)
2. Run all the scripts. See that they don't produce errors.

### CODE REVIEW

**Comment by Szeket** [comment]
Build is significantly faster! But then the page loads for a long time. And some words search slower, mostly pictures are loading longer than on live version. Everything else seems OK to me :)

**Comment by femans** [comment]
@dokterbob note that there is 1 breaking change, described in the readme.md, the nsfw api env variable is different. 

You might also want to check that the CI picks up the test scripts properly.

**Review by dokterbob — CHANGES_REQUESTED** [review]
# Resolved by fixup
## Tests are failing
image

## Serve is failing
image

## Fixup
Installing `sass` seems to resolve aforementioned issues.
```shell
% pnpm add -D sass
```

# Unresolved

## Lint is failing
```shell
% pnpm run-script lint 

> dweb-search-frontend@0.1.0 lint /Users/drbob/Development/dweb-search-frontend
> eslint --ext .js,.mjs,.cjs,.jsx,.ts,.tsx,.vue src


Oops! Something went wrong! :(

ESLint: 8.17.0

ESLint couldn't find the plugin "eslint-plugin-jsx-a11y".

(The package "eslint-plugin-jsx-a11y" was not found when loaded as a Node module from the directory "/Users/drbob/Development/dweb-search-frontend".)

It's likely that the plugin isn't installed correctly. Try reinstalling by running the following:

    npm install eslint-plugin-jsx-a11y@latest --save-dev

The plugin "eslint-plugin-jsx-a11y" was referenced from the config file in ".eslintrc.js » eslint-config-airbnb » /Users/drbob/Development/dweb-search-frontend/node_modules/.pnpm/eslint-config-airbnb@19.0.4_3yxiwxzsqipdmy4jwrlv6vgfmy/node_modules/eslint-config-airbnb/rules/react-a11y.js".

If you still can't figure out the problem, please stop by https://eslint.org/chat/help to chat with the team.

 ELIFECYCLE  Command failed with exit code 2.
```

It seems that we're trying to use the airbnb config, which includes all kinds of React cruft as peer dependencies, which we don't need here. I think we should have this: https://www.npmjs.com/package/eslint-config-airbnb-base

Replacing the airbnb config and consequently extending config from `airbnb-base` gets the linter to run.
```shell
pnpm uninstall -D eslint-config-airbnb && pnpm install -D eslint-config-airbnb-base@latest
```

However, it now gives the following errors:
```shell
% pnpm lint

> dweb-search-frontend@0.1.0 lint /Users/drbob/Development/dweb-search-frontend
> eslint --ext .js,.mjs,.cjs,.jsx,.ts,.tsx,.vue src


Error while parsing /Users/drbob/Development/dweb-search-frontend/src/plugins/audioPlugin.js
Line 10, column 11: Unexpected token =
`parseForESLint` from parser `/Users/drbob/Development/dweb-search-frontend/node_modules/.pnpm/vue-eslint-parser@8.3.0_eslint@8.17.0/node_modules/vue-eslint-parser/index.js` is invalid and will just be ignored

Error while parsing /Users/drbob/Development/dweb-search-frontend/src/helpers/FetchDoggy.js
Line 28, column 11: Unexpected token ;
`parseForESLint` from parser `/Users/drbob/Development/dweb-search-frontend/node_modules/.pnpm/vue-eslint-parser@8.3.0_eslint@8.17.0/node_modules/vue-eslint-parser/index.js` is invalid and will just be ignored

/Users/drbob/Development/dweb-search-frontend/src/components/results/detail/helpers/DocumentViewer.vue
  38:23  error  Parse errors in imported module '@/helpers/FetchDoggy': Unexpected token ; (28:11)  import/no-named-as-default
  38:23  error  Parse errors in imported module '@/helpers/FetchDoggy': Unexpected token ; (28:11)  import/no-named-as-default-member

/Users/drbob/Development/dweb-search-frontend/src/helpers/FetchDoggy.js
  28:11  error  Parsing error: Unexpected token ;

/Users/drbob/Development/dweb-search-frontend/src/plugins/audioPlugin.js
  10:11  error  Parsing error: Unexpected token =

✖ 4 problems (4 errors, 0 warnings)

 ELIFECYCLE  Command failed with exit code 1.
```

## Results not loading
I'm getting a `TypeError: url.parse is not a function` on Brave V1.39.111.
image



## Preview is failing
```shell
% pnpm run-script preview

> dweb-search-frontend@0.1.0 preview /Users/drbob/Development/dweb-search-frontend
> vite preview --open --port 8080

  > Local: http://localhost:8080/
  > Network: use `--host` to expose
```
image

## Fixups suggestions
https://github.com/ipfs-search/dweb-search-frontend/compare/vite-migration...fixups_vite_mathijs?expand=1

File: README.md [hunk 1]
```diff
@@ -38,9 +38,9 @@ See [Configuration Reference](https://cli.vuejs.org/config/).
 This build uses an API to check nsfw content. 
 
 The default API endpoint to is: https://api.ipfs-search.com/nsfw/
-. This can be overridden by injecting `process.env.NSFW_API` 
+. This can be overridden by injecting environment variable `VITE_NSFW_API` 
 
-The API call should be: `<process.env.NSFW_API><CID>`, so e.g.
+The API call should be: `<VITE_NSFW_API><CID>`, so e.g.
```
[the remark this answers is not in the document]
Comment (femans) [reply README.md:43]: Vite wants to protect you against exposing database passwords in the code. This is why anything that is not prefixed with VITE_ is not available in the frontend code.

File: package.json [hunk 2]
```diff
@@ -4,55 +4,53 @@
...
-    "lint": "vue-cli-service lint",
-    "dev": "vue-cli-service serve --skip-plugins @vue/cli-plugin-eslint"
+    "serve": "vite --open",
+    "preview": "vite preview --open --port 8080",
+    "build": "vite build --out-dir dist",
+    "lint": "eslint --ext .js,.mjs,.cjs,.jsx,.ts,.tsx,.vue src",
+    "lint:fix": "eslint --ext .js,.mjs,.cjs,.jsx,.ts,.tsx,.vue src --fix",
+    "test": "vitest watch",
+    "test:once": "vitest run",
+    "test:coverage": "vitest run --coverage"
```
[the remark this answers is not in the document]
Comment (femans) [reply package.json:14]: they seem pretty obvious to me.

**Comment by femans** [comment]
> # Resolved by fixup
> ## Tests are failing
> image
> 
> ## Serve is failing
> image
> 
> ## Fixup
> Installing `sass` seems to resolve aforementioned issues.
> 
> ```shell
> % pnpm add -D sass
> ```
> 
> # Unresolved
> ## Lint is failing
> ```shell
> % pnpm run-script lint 
> 
> > dweb-search-frontend@0.1.0 lint /Users/drbob/Development/dweb-search-frontend
> > eslint --ext .js,.mjs,.cjs,.jsx,.ts,.tsx,.vue src
> 
> 
> Oops! Something went wrong! :(
> 
> ESLint: 8.17.0
> 
> ESLint couldn't find the plugin "eslint-plugin-jsx-a11y".
> 
> (The package "eslint-plugin-jsx-a11y" was not found when loaded as a Node module from the directory "/Users/drbob/Development/dweb-search-frontend".)
> 
> It's likely that the plugin isn't installed correctly. Try reinstalling by running the following:
> 
>     npm install eslint-plugin-jsx-a11y@latest --save-dev
> 
> The plugin "eslint-plugin-jsx-a11y" was referenced from the config file in ".eslintrc.js » eslint-config-airbnb » /Users/drbob/Development/dweb-search-frontend/node_modules/.pnpm/eslint-config-airbnb@19.0.4_3yxiwxzsqipdmy4jwrlv6vgfmy/node_modules/eslint-config-airbnb/rules/react-a11y.js".
> 
> If you still can't figure out the problem, please stop by https://eslint.org/chat/help to chat with the team.
> 
>  ELIFECYCLE  Command failed with exit code 2.
> ```
> 
> It seems that we're trying to use the airbnb config, which includes all kinds of React cruft as peer dependencies, which we don't need here. I think we should have this: https://www.npmjs.com/package/eslint-config-airbnb-base
> 
> Replacing the airbnb config and consequently extending config from `airbnb-base` gets the linter to run.
> 
> ```shell
> pnpm uninstall -D eslint-config-airbnb && pnpm install -D eslint-config-airbnb-base@latest
> ```
> 
> However, it now gives the following errors:
> 
> ```shell
> % pnpm lint
> 
> > dweb-search-frontend@0.1.0 lint /Users/drbob/Development/dweb-search-frontend
> > eslint --ext .js,.mjs,.cjs,.jsx,.ts,.tsx,.vue src
> 
> 
> Error while parsing /Users/drbob/Development/dweb-search-frontend/src/plugins/audioPlugin.js
> Line 10, column 11: Unexpected token =
> `parseForESLint` from parser `/Users/drbob/Development/dweb-search-frontend/node_modules/.pnpm/vue-eslint-parser@8.3.0_eslint@8.17.0/node_modules/vue-eslint-parser/index.js` is invalid and will just be ignored
> 
> Error while parsing /Users/drbob/Development/dweb-search-frontend/src/helpers/FetchDoggy.js
> Line 28, column 11: Unexpected token ;
> `parseForESLint` from parser `/Users/drbob/Development/dweb-search-frontend/node_modules/.pnpm/vue-eslint-parser@8.3.0_eslint@8.17.0/node_modules/vue-eslint-parser/index.js` is invalid and will just be ignored
> 
> /Users/drbob/Development/dweb-search-frontend/src/components/results/detail/helpers/DocumentViewer.vue
>   38:23  error  Parse errors in imported module '@/helpers/FetchDoggy': Unexpected token ; (28:11)  import/no-named-as-default
>   38:23  error  Parse errors in imported module '@/helpers/FetchDoggy': Unexpected token ; (28:11)  import/no-named-as-default-member
> 
> /Users/drbob/Development/dweb-search-frontend/src/helpers/FetchDoggy.js
>   28:11  error  Parsing error: Unexpected token ;
> 
> /Users/drbob/Development/dweb-search-frontend/src/plugins/audioPlugin.js
>   10:11  error  Parsing error: Unexpected token =
> 
> ✖ 4 problems (4 errors, 0 warnings)
> 
>  ELIFECYCLE  Command failed with exit code 1.
> ```
> 
> ## Results not loading
> I'm getting a `TypeError: url.parse is not a function` on Brave V1.39.111. image
> 
> ## Preview is failing
> ```shell
> % pnpm run-script preview
> 
> > dweb-search-frontend@0.1.0 preview /Users/drbob/Development/dweb-search-frontend
> > vite preview --open --port 8080
> 
>   > Local: http://localhost:8080/
>   > Network: use `--host` to expose
> ```
> 
> image
> 
> ## Fixups suggestions
> https://github.com/ipfs-search/dweb-search-frontend/compare/vite-migration...fixups_vite_mathijs?expand=1

Are you sure you ran `npm install` beforehand? Because vite has built-in sass support. Moreover, the fleek build works fine.

File: package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (dokterbob) [reply package.json:14]: That's great. You're a very experienced developer.
We would love contributions from those who are not though.

Plus, I'm an experienced developer and dunno how preview is different from serve.

**Review by dokterbob — CHANGES_REQUESTED** [review]
# Requested changes
- [ ] Resolve linter issues.
- [ ] Either remove `preview` command (what is it even for?) or resolve issues there.
- [ ] Document available commands in README or link from README to a place where they are documented.
- [ ] https://github.com/ipfs-search/dweb-search-frontend/pull/205/files#r893490553

# Suggested changes
I would really appreciate this, I think it will benefit us in the future, but it's no hard requirement. However, I do not think we will 'fix this later', unless maybe we create issues for them. So, if you think it's a desirable thing, please either fix it or create an issue (referring this PR) before we're merged.
- [ ] https://github.com/ipfs-search/dweb-search-frontend/pull/205/files#r893485526
- [ ] If you know this is positive, would be great to have a ticket. If not, let's leave it. https://github.com/ipfs-search/dweb-search-frontend/pull/205/files#r893479273
- [ ] https://github.com/ipfs-search/dweb-search-frontend/pull/205/files#r893487610

# Testing Fleek/deploy
It works!

# Testing commands
This time based on `npm` over `pnpm`. I've removed `node_modules` and started with `npm i`.

```shell
% npm --version
8.12.1
% node --version
v16.15.0
```

## `npm run lint`
```
> dweb-search-frontend@0.1.0 lint
> eslint --ext .js,.mjs,.cjs,.jsx,.ts,.tsx,.vue src

Warning: React version was set to "detect" in eslint-plugin-react settings, but the "react" package is not installed. Assuming latest React version for linting.

Error while parsing /Users/drbob/Development/dweb-search-frontend/src/plugins/audioPlugin.js
Line 10, column 11: Unexpected token =
`parseForESLint` from parser `/Users/drbob/Development/dweb-search-frontend/node_modules/vue-eslint-parser/index.js` is invalid and will just be ignored

Error while parsing /Users/drbob/Development/dweb-search-frontend/src/helpers/FetchDoggy.js
Line 28, column 11: Unexpected token ;
`parseForESLint` from parser `/Users/drbob/Development/dweb-search-frontend/node_modules/vue-eslint-parser/index.js` is invalid and will just be ignored

/Users/drbob/Development/dweb-search-frontend/src/components/results/detail/helpers/DocumentViewer.vue
  38:23  error  Parse errors in imported module '@/helpers/FetchDoggy': Unexpected token ; (28:11)  import/no-named-as-default
  38:23  error  Parse errors in imported module '@/helpers/FetchDoggy': Unexpected token ; (28:11)  import/no-named-as-default-member

/Users/drbob/Development/dweb-search-frontend/src/helpers/FetchDoggy.js
  28:11  error  Parsing error: Unexpected token ;

/Users/drbob/Development/dweb-search-frontend/src/plugins/audioPlugin.js
  10:11  error  Parsing error: Unexpected token =

✖ 4 problems (4 errors, 0 warnings)
 ```

I also tested with the `.nvmrc`-mandated versions (node v16.15.1, npm v8.11.0) and ran into the same issue.

## `npm run test`
Passing.

## `npm run preview`
Opens the browser, but no results are loading, same error as before.
image

## `npm run build`
Build succeeds, the problem found in preview does not seem to occur here. I did note that resources are much better grouped and so the initial page load should be way faster!

## `npm run serve`
Working just fine, the error found in preview hasn't popped up here.

## `npm test:coverage`
VERY sweet!

**Pull request closed without merging**